core/src/main/scala/org/bykn/bosatsu/rankn/TypeEnv.scala

```diff
@@ -37,26 +35,50 @@ class TypeEnv[+A] private (
   def allDefinedTypes: List[DefinedType[A]] =
     definedTypes.values.toList.sortBy { dt => (dt.packageName, dt.name) }
 
-  def getConstructor(p: PackageName, c: Constructor): Option[(List[(Bindable, Type)], DefinedType[A], Type)] =
+  def getConstructor(p: PackageName, c: Constructor): Option[(DefinedType[A], ConstructorFn)] =
     constructors.get((p, c))
 
+  def getConstructorParams(p: PackageName, c: Constructor): Option[List[(Bindable, Type)]] =
+    constructors.get((p, c)).map(_._2.args)
+
   def getType(p: PackageName, t: TypeName): Option[DefinedType[A]] =
     definedTypes.get((p, t))
 
-  def getValue(p: PackageName, n: Identifier): Option[Type] =
+  def getExternalValue(p: PackageName, n: Identifier): Option[Type] =
     values.get((p, n))
 
-  def localValuesOf(p: PackageName): SortedMap[Identifier, Type] =
-    (SortedMap.newBuilder[Identifier, Type] ++= values.iterator.collect { case ((pn, n), v) if pn == p => (n, v) }).result()
+  private def getConsType[A1](dt: DefinedType[A1], cf: ConstructorFn)(implicit ev: A1 <:< Variance): Type =
```

Comment:
we don't really use `ev` currently, but it is here to enforce that we only access the constructor type *after* we have inferred the types (or variance) of the defined types.